### PR TITLE
report bundle status as prometheus metrics

### LINKED ISSUE #1472: OPA Status - expose as prometheus metrics
# Expected Behavior
Bundle information that exposed via status API should be exposed also as Prometheus metrics (last bundle load time, last bundle version, etc) 

# Actual Behavior
Exposed only by implementing status API

If this makes sense, I might be able to provide a PR.

### PR DESCRIPTION
**Pull request by omerlh**
Signed-off-by: omerlh <omerl@soluto.com>

close #1472 

I added a simple metric, `bundle_status_phase` with four labels. 
For example: 
```
bundle_status_phase{name="discovery/config",phase="download",revision="",status="failure"} 1
```
I wanted to get feedback before adding documentation.

### CODE REVIEW

File: plugins/bundle/status.go [hunk 1]
```diff
@@ -27,30 +29,42 @@ type Status struct {
...
+func (s *Status) SetError(err error, phase string) {
+	
 	if err == nil {
 		s.Code = ""
 		s.Message = ""
 		s.Errors = nil
 		return
 	}
 
+	bundleStatusPhase.WithLabelValues(phase, "failure", "", s.Name).Set(1)
```
Comment (patrick-east) [inline plugins/bundle/status.go]: Going off the docstring for `SetError` should we clear the failure status on the gauge label if `err == nil` too?

File: plugins/bundle/plugin.go [hunk 2]
```diff
@@ -144,23 +146,23 @@ func (p *Plugin) oneShot(ctx context.Context, u download.Update) {
...
 	}
 
 	if u.Bundle != nil {
+		bundleStatusPhase.WithLabelValues("activate", "success", "").Set(1)
 		p.status.SetDownloadSuccess()
 
 		if err := p.activate(ctx, u.Bundle); err != nil {
 			p.logError("Bundle activation failed: %v", err)
-			p.status.SetError(err)
+			p.status.SetError(err, "activate")
```
Comment (patrick-east) [inline plugins/bundle/plugin.go]: Can we make these more of an enum/const set of possible statuses?

**Comment by omerlh** [comment]
I did some refactor to make metrics collections more reliable. Now the metrics are collected using a custom collector, that can collect the current status. With that change, I moved to expose 2 metrics:
```
# HELP bundle_last_successful_activation The time of the last succesful activation
# TYPE bundle_last_successful_activation gauge
bundle_last_successful_activation{name="discovery/config"} -6.21355968e+10
# HELP bundle_last_successful_download The time of the last succesful download
# TYPE bundle_last_successful_download gauge
bundle_last_successful_download{name="discovery/config"} -6.21355968e+10
```

I tried also to expose a metric for failure - but for some reason, when the discovery download failed, there was no indication on the status - it looks like `Message`, `Code` and `Errors` are empty or null. Anyway, I figured that it best to write alert using the timers - e.g. something like 
```
time() - bundle_last_successful_download > 1h
```
The promql syntax is not perfect, but this is the general idea. This will allow you to monitor that everything works as expected :)

About the revision - I was wondering if it's a good idea to add the last revision as metrics. As it's a string, I can only put it on a label. Because high cardinality label is not recommended (see [here](https://prometheus.io/docs/practices/naming/)), I wasn't sure it's a good idea to put it as a label. This could lead to serious issues if the revision changed frequently. The best solution is putting it as the metric value (like I did with the time), but this is possible only if the revision is number.

To conclude, I think this is good enough - and this, combined with console status API (similar to #1490) will give good coverage for bundle issues.

One last thing - about the status phases I added to the `SetError` - I don't need it anymore, before deleting it - do you think this is something worth keeping?

**Comment by patrick-east** [comment]
I was chatting with @tsandall about this a little bit and some concerns arose around adding the prometheus dependency on the bundle & discovery plugins. We already have it in the server, and there is actually some feedback that we should have it be entirely optional. The bloat it adds to consumers of OPA as a golang library is pretty significant.

I'm also a little bit concerned that we are forcing some monitoring aspects onto bundle status through prometheus that isn't really a natural fit. The API's and data provided is kind of an awkward form this way. I'm concerned that this is not as generally consumable as it should be for being a part of the core OPA library+service.

One more thing to clarify:
>To conclude, I think this is good enough - and this, combined with console status API (similar to #1490) will give good coverage for bundle issues.

The changes in #1490 are only for decision logs, not for the status logs.

Brainstorming a little bit, here a couple of things I can think of that might help with the issues:

* We split out the prometheus bits for bundle monitoring to be a plugin fully external to OPA. In theory the plugin can register a listener for bundle status updates and perform all of this sort of thing while keeping the dependencies out and making it entirely optional. We could maybe even add a `plugin` config option to the decision logger and define a plugin API specific to status service plugins.

* We add in a `/status` endpoint to OPA that exposes the current status in a similar format to what is pushed to services that implement the Status API. If I put on my sysadmin hat and look at the problem of monitoring OPA this would likely be much more useful to me. I could hook up nagios monitors, or custom scripts/tools that poll the API and export into any TSDB (prometheus included).

* We do some updating to the logs from OPA with regards to the bundle download status that is more easily consumed by a logging stack that could be used to set monitors and alarm on errors. I haven't looked through them much but I'm guessing there are some improvements we can make for this.

@omerlh @tsandall thoughts?

**Comment by omerlh** [comment]
This sounds pretty good. If you look at my code, I added a collector - this collector can be easily extracted to a plugin - just needs some wiring. I really like the direction you're proposing. 
Regarding logging - I know that the PR was only for decision log, I was pointing out that having the same functionality for status API will be great. I think this is what you referred to in your last comment. Looking forward to seeing this! I'll keep the branch for now, just in case the code will be relevant at some point,

**Comment by omerlh** [comment]
It's been a while - are there any plans to add one of the futures describe here in the near future? I think the easiest will be to add the revision to the logs, this will kinda cover what we needs.

**Comment by tsandall** [comment]
Including the currently active bundle revision in the bundle plugin logs seems like a relatively easy change. WDYT @patrick-east ?

**Pull request closed without merging**